### PR TITLE
Create tutorial-01.md file

### PR DESCRIPTION
**Pull request by ipbabble**
This file contains a basic introductory tutorial on buildah. It covers 
buildah from a base
briefly describes containers/storage and containers/image
buildah from scratch
installing packages and files to a scratch image 
buildah push and running a buildah built container in docker
buildah bud

### CODE REVIEW

File: docs/tutorial-01.md [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+
+    dnf install --installroot $scratchmnt --release 26 bash coreutils --setopt install_weak_deps=false -y
+
+Let's try it out:
+
+    # buildah run $newcontainer bash
+    bash-4.4# cd /usr/bin
+    bash-4.4# exit
+
+Notice we have a `/usr/bin` directoy in the newcontianers image layer. Let's first copy a simple file. Create a file called runecho.sh contains the following:
```
Comment (twaugh) [inline docs/tutorial-01.md]: typo: newcontainers

File: docs/tutorial-01.md [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+    buildah inspect --type=image fedora-bashecho
+
+## OCI images built using buildah are portable
+
+Let's test if this new OCI image is really portable to another OCI technology like docker. First you shouuld install docker and start it.
+
+    dnf -y install docker
+    systemctl start docker
+    
+The we need to push the image from the `containers/image` across to docker's storage area which is under `/var/lib/docker`. This is managed by the docker dameon. We can do this telling buildah to push to the docker protocol called docker-daemon.
```
Comment (twaugh) [inline docs/tutorial-01.md]: typo: daemon

File: docs/tutorial-01.md [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+
+    buildah push fedora-bashecho docker-daemon:fedora-bashecho:latest
+
+Underneath `buildah` the `containers/storage` copies the image blob over to the docker storage. This can take a little while. And usually you won't need to do this. If you're using `buildah` you are probably not using docker. This is just for demo purposes. Letr's try it:
```
Comment (twaugh) [inline docs/tutorial-01.md]: typo: Let's

File: docs/tutorial-01.md [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+    RUN echo "Updating all fedora packages"; dnf -y update; dnf -y clean all
+    RUN echo "Installing httpd"; dnf -y install httpd
+ 
+    # Expose the default httpd port 80
+    EXPOSE 80
+
+    # Run the httpd
+    CMD ["/usr/sbin/httpd", "-DFOREGROUND"]
+
+Now run `buildah bud` with the name of the Dockerfile and the image name (fedora-hhtpd):
```
Comment (twaugh) [inline docs/tutorial-01.md]: typo: fedora-httpd

File: docs/tutorial-01.md [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+
+    buildah bud -f Dockerfile -t fedora-httpd
+
+You will see all the steps of the Dockerfile executing. Afterware `buildah images` will show you the new image. Now we need to create the container and test it with `buidlah run` it:
```
Comment (twaugh) [inline docs/tutorial-01.md]: typo: afterwards

File: docs/tutorial-01.md [hunk 6]
```diff
@@ -0,0 +1,210 @@
...
+
+What if you have been using docker for while and have some exisiting Dockerfiles. Not a problem. `buildah` can build images using a Dockerfile. The `build-using-dockerfile`, or `bud` for short, takes a Dockefile as input and produces an OCI image.
+
+Find one of your Dockerfiles or create a file called Dockerfile. Use the following example or some variation if you'd like:
+
+    # Base on the Fedora
+    FROM fedora:latest
+    MAINTAINER W Henry email ipbabble@gmail.com
+
+    # Update image and install hhtpd
```
Comment (twaugh) [inline docs/tutorial-01.md]: typo: httpd

File: docs/tutorial-01.md [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+    This is a new container from ipbabble [ 8 ]
+    This is a new container from ipbabble [ 9 ]
+    
+OCI container images built with `buildah` are completely standard as expected. So now it might be time to run:
+
+    dnf -y remove docker
+
+## Using Dockerfiles with buildah
+
+What if you have been using docker for while and have some exisiting Dockerfiles. Not a problem. `buildah` can build images using a Dockerfile. The `build-using-dockerfile`, or `bud` for short, takes a Dockefile as input and produces an OCI image.
```
Comment (twaugh) [inline docs/tutorial-01.md]: typo: existing

File: docs/tutorial-01.md [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+
+We can also see that there are also no containers by running:
+
+    buildah containers
+  
+When you build a working container from an existing image, buildah defaults to appending '-working-container' to the image's name to construct a name for the container. The buildah CLI conveniently returns the name of the new container. You can take advantage of this by assigning the returned value to a shell varible using standard shell assignment :
+
+    container=$(buildah from fedora)
+
+Of course you can always just run `build from fedora` and then list the containers to see the contianer name. As mentioned previously it will be fedora-working-container in this case.
```
Comment (twaugh) [inline docs/tutorial-01.md]: typo: container

File: docs/tutorial-01.md [hunk 9]
```diff
@@ -0,0 +1,210 @@
+# buildah Tutorial 1
+## Building OCI container images
+
+The purpose of this tutorial is to deomonstrate how buildah can be used to build container images compliant with the [Open Container Initiative](https://www.opencontainers.org/) (OCI) [image specification](https://github.com/opencontainers/image-spec). Images can be built from existing images, from scratch, and using Dockerfiles. OCI images built using the buildah command line tool (CLI) and the underlying OCI based technologies (e.g. [containers/image](https://github.com/containers/image) and [containers/storage](https://github.com/containers/storage)) are portable and can therefore run in a docker environment.
+
+In brief the `containers/image` project provides mecahnisms to copy, push, pull, inspect and sign container images. The `containers/storage` project provides mecahnisms for storing filesystem layers, container images, and containers. `buildah` is a CLI that takes advantage of these underlying projects and therefore allows you to build, move, and manage container images and containers.
```
Comment (twaugh) [inline docs/tutorial-01.md]: typo: mechanisms (x 2 )

File: docs/tutorial-01.md [hunk 10]
```diff
@@ -0,0 +1,210 @@
+# buildah Tutorial 1
+## Building OCI container images
+
+The purpose of this tutorial is to deomonstrate how buildah can be used to build container images compliant with the [Open Container Initiative](https://www.opencontainers.org/) (OCI) [image specification](https://github.com/opencontainers/image-spec). Images can be built from existing images, from scratch, and using Dockerfiles. OCI images built using the buildah command line tool (CLI) and the underlying OCI based technologies (e.g. [containers/image](https://github.com/containers/image) and [containers/storage](https://github.com/containers/storage)) are portable and can therefore run in a docker environment.
```
Comment (twaugh) [inline docs/tutorial-01.md]: typo: demonstrate

**Comment by TomSweeneyRedHat** [comment]
@ipbabble this is some serious awesome-sauce!  I just took a quick scan, it looks great overall.  I spotted a couple of nits, will go through in more detail later today.  We should spin one of these up for the kpod tools too, perhaps skopeo.

**Comment by rhatdan** [comment]
Lets turn it into a blog as well.

File: docs/tutorial-01.md (hunk 8 above)
Comment (nalind) [reply docs/tutorial-01.md]: typo: "build from fedora" should be "buildah from fedora"

File: docs/tutorial-01.md [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+
+See the name:  
+
+    echo $container
+
+What can we do with this new container? Let's try running bash:
+
+    buildah run $container bash
+    
+Notice we get a new shell prompt because we running a bash shell inside of the container. It should be noted that `buildah run` is not intended for running production containers. It is for helping debug during the build process.
```
Comment (nalind) [inline docs/tutorial-01.md]: Needs an "are" between "we" and "running".

File: docs/tutorial-01.md [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+
+    container_linux.go:274: starting container process caused "exec: \"java\": executable file not found in $PATH"
+    
+Lets try installing it using:
+    
+    buildah run $container -- dnf install java
+
+The `--` syntax basically says: no more `buildah run` command options after this point. It is required if the command we specify includes command line options which are not meant for buildah. 
+
+Now running `java` will show that Java has been installed. It will return the `Usage`:
```
Comment (nalind) [inline docs/tutorial-01.md]: I think either "\`java\` in the container" or "\`buildah run java\`" would be clearer here.

**Comment by ipbabble** [comment]
Thanks for all the feedback. I forgot the DCO on the commit so I have to
recommit before this PR will be acceptable. I'll make these changes before
I do the commit and PR.

William

William Henry
Senior Consulting Software Engineer
Red Hat
Colorado, USA.

File: docs/tutorial-01.md [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+  
+You can see this new empty container by running:
+
+    buildah containers
+  
+Its container name is working-container by default. And it's stored in the `$newcontainer` variable. Notice the image name is scratch. And when we run:
+
+    buildah images
+  
+We don't see it listed. That is because "scratch" is not an image. It is an empty container.
```
Comment (nalind) [inline docs/tutorial-01.md]: "scratch" won't show up in the containers list, either, and commands that deal with containers won't see it.  It's just a name that we specifically recognize as meaning "no image".  I'm harping on this, probably a little too much, because I think it's very important that we are always clear about whether we're talking about containers or images, so that readers don't get confused.

File: docs/tutorial-01.md (hunk 1 above)
Comment (nalind) [reply docs/tutorial-01.md]: typo: "directoy" should be "directory"

File: docs/tutorial-01.md [hunk 14]
```diff
@@ -0,0 +1,210 @@
...
+    This is a new container from ipbabble [ 5 ]
+    This is a new container from ipbabble [ 6 ]
+    This is a new container from ipbabble [ 7 ]
+    This is a new container from ipbabble [ 8 ]
+    This is a new container from ipbabble [ 9 ]
+
+It works! Congratulations, you have built a new OCI container from scratch that uses bash scripting. Let's add some more configuration information.
+
+    buildah config --created-by "ipbabble"  $newcontainer
+    buildah config --author "wgh at redhat.com @ipbabble" --label fedora26-bashecho:latest $newcontainer
```
Comment (nalind) [inline docs/tutorial-01.md]: I think that --label looks like an image name with a tag, which could get confusing when you commit it using a similar name below.  Maybe a different example, using the `name=value` syntax to specify the label's name and its value?

File: docs/tutorial-01.md (hunk 2 above)
Comment (nalind) [reply docs/tutorial-01.md]: Maybe "Let's copy that image from where and how containers/storage stores it to where and how the docker daemon stores its images, so that we can run it using docker."?

File: docs/tutorial-01.md (hunk 3 above)
Comment (nalind) [reply docs/tutorial-01.md]: Maybe "Under the covers, the containers/image library calls into the containers/storage library to read the image's contents, and sends them to the local docker daemon."?

**Comment by ipbabble** [comment]
Nalin,

"scratch" absolutely shows up in the containers list. It shows up as the
IMAGE NAME of the working-container.

ac8fa6be0f0a     *                  scratch
working-container


William

William Henry
Senior Consulting Software Engineer
Red Hat
Colorado, USA.

File: docs/tutorial-01.md (hunk 5 above)
Comment (nalind) [reply docs/tutorial-01.md]: typo: "buidlah" should be "buildah"

File: docs/tutorial-01.md (hunk 13 above)
Comment (ipbabble) [reply docs/tutorial-01.md]: scratch does show up. as the IMAGE NAME of working-container.

ac8fa6be0f0a     *                  scratch                          working-container

**Comment by nalind** [comment]
You're right, the placeholder value shows up as the name of the image which was used to create the container, so that we always have a value there.  The wording suggested to me that we'd expect to see a container named "scratch" in the list, so maybe rewording it would make that clearer.

**Comment by nalind** [comment]
Can you sign the commits?  There's a longer explanation in the new CONTRIBUTING.md, but it boils down to adding a "Signed-off-by: William Henry <emailaddress>" to the commit log messages for each patch in the PR.  That looks like the reason the CI tests aren't passing.

**Comment by nalind** [comment]
This is a good read.  Thanks for putting it together.

File: docs/tutorial-01.md (hunk 10 above)
Comment (ipbabble) [reply docs/tutorial-01.md]: ack done (will recommit when all comments are acked and I figure out /remember how to do it ;-)

File: docs/tutorial-01.md (hunk 9 above)
Comment (ipbabble) [reply docs/tutorial-01.md]: ack done

File: docs/tutorial-01.md (hunk 8 above)
Comment (ipbabble) [reply docs/tutorial-01.md]: ack done

File: docs/tutorial-01.md (hunk 1 above)
Comment (ipbabble) [reply docs/tutorial-01.md]: s/newcontainers/newcontainer's/
ack done

File: docs/tutorial-01.md (hunk 2 above)
Comment (ipbabble) [reply docs/tutorial-01.md]: Going with:
"Let's copy that image from where and how containers/storage stores it to where and how the docker daemon stores its images, so that we can run it using docker. We can achieve this using `buildah push`. This copies the image to docker's repository area which is located under `/var/lib/docker`. Docker's repository is managed by the docker daemon. This needs to be explicitly stated by telling buildah to push to the docker repository protocol using `docker-daemon:`."

File: docs/tutorial-01.md (hunk 3 above)
Comment (ipbabble) [reply docs/tutorial-01.md]: ack done.

File: docs/tutorial-01.md (hunk 7 above)
Comment (ipbabble) [reply docs/tutorial-01.md]: ack done

File: docs/tutorial-01.md (hunk 6 above)
Comment (ipbabble) [reply docs/tutorial-01.md]: I noticed last night that my fingers kept doing this. I guess I didn't fix enough of them. 
ack done

File: docs/tutorial-01.md (hunk 4 above)
Comment (ipbabble) [reply docs/tutorial-01.md]: ack done

File: docs/tutorial-01.md (hunk 5 above)
Comment (ipbabble) [reply docs/tutorial-01.md]: ack done

File: docs/tutorial-01.md (hunk 11 above)
Comment (ipbabble) [reply docs/tutorial-01.md]: Good catch. I read it again and my mind inputed the "are". Stop playing tricks Mind!
Done.

File: docs/tutorial-01.md (hunk 12 above)
Comment (ipbabble) [reply docs/tutorial-01.md]: Done. `buildah run $container java`

File: docs/tutorial-01.md (hunk 14 above)
Comment (ipbabble) [reply docs/tutorial-01.md]: You are right and it was something I noticed but didn't know what I was really doing here. And it was confusing. 
name=fedora26-bashecho 
is better
Or should I use a different label completely? I want to show usage of --label 
any thoughts??

File: docs/tutorial-01.md (hunk 13 above)
Comment (ipbabble) [reply docs/tutorial-01.md]: Does this help?

Its container name is working-container by default. And it's stored in the `$newcontainer` variable. Notice the image name (IMAGE NAME) is "scratch". This just indicates that there is no real image yet. i.e. It is containers/storage but there is no representation in containers/image. So when we run:

    buildah images
  
We don't see the image listed. That is because "scratch" is not an image. It is an empty container.

Comment (ipbabble) [reply docs/tutorial-01.md]: Nope. better:
We don't see the image listed. There is no image. It is an empty container.

**Comment by ipbabble** [comment]
I guess I close this one and then backout my commit on my fork and start over with the DCO and a new PR. Right?

**Comment by TomSweeneyRedHat** [comment]
Assuming you've gitconfig setup with your username, I think you can just do changes, "git add {files}", "git commit --amend -s" and then push it over again.

**Comment by ipbabble** [comment]
<eyeroll>

**Comment by nalind** [comment]
With multiple commits, 'git rebase -i <upstreamreponame>/master', then replacing 'pick' with 'reword' in the list of commits that it initially offers will let you edit the commit logs for each patch in turn, which should let you append the Signed-off-by: that the validation test expects.

File: docs/tutorial-01.md (hunk 14 above)
Comment (nalind) [reply docs/tutorial-01.md]: I had to go looking.  The fedora:latest image includes these examples:
* license=MIT
* name=Fedora
* vendor="Fedora Project"
* version=26

File: docs/tutorial-01.md (hunk 13 above)
Comment (nalind) [reply docs/tutorial-01.md]: Ah, the "it" is what confused me, because the previous sentence was talking about images.  Maybe "The container has not contents."?
